### PR TITLE
Fix/sonarjnkns 256: Support credentials plugin

### PR DESCRIPTION
**Pull request by luka5**
This fixes the [SONARJNKNS-256](https://jira.sonarsource.com/browse/SONARJNKNS-256). We use the Jenkins credentials plugin instead of saving the token directly inside the xml file. The tests are extended, updated and running. A manually created `hpi` file works as expected.

The changes are not compatible to the previous version, since the former `serverAuthenticationToken` gets removed completely. Is there a way to provide a migration strategy? We think, creating credentials automatically is not a proper way to handle that.

image

`hudson.plugins.sonar.SonarGlobalConfiguration.xml`:
```xml
<?xml version='1.0' encoding='UTF-8'?>
<hudson.plugins.sonar.SonarGlobalConfiguration plugin="sonar@2.X">
  <installations>
    <hudson.plugins.sonar.SonarInstallation>
      <name>SonarQube Server</name>
      <serverUrl></serverUrl>
      <credentialsId>sonarqube_token</credentialsId>
....
```

What's your opinion on these changes? Do you want us to change something? We are already using it and it works exactly the way we want.

### CODE REVIEW

File: src/main/java/hudson/plugins/sonar/SonarInstallation.java [hunk 1]
```diff
@@ -37,9 +41,9 @@
   private final String serverUrl;
 
   /**
-   * @since 2.4
+   * @since X
```
Comment (henryju) [inline src/main/java/hudson/plugins/sonar/SonarInstallation.java]: To be fixed

File: src/main/java/hudson/plugins/sonar/SonarInstallation.java [hunk 2]
```diff
@@ -135,12 +139,25 @@ public String getServerUrl() {
...
+    if (cred == null) { return null; }
+    return ((PasswordCredentials) cred).getPassword().getPlainText();
   }
 
+  public StandardCredentials getCredentials(Run<?, ?> build) {
+    return CredentialsProvider.findCredentialById(credentialsId, StandardCredentials.class, build);
+  }
+
+  /**
+   * @since X
```
Comment (henryju) [inline src/main/java/hudson/plugins/sonar/SonarInstallation.java]: TBD

**Comment by henryju** [comment]
@drautureau-sonarsource  What you you think of the proposed feature?

File: src/main/java/hudson/plugins/sonar/SonarInstallation.java [hunk 3]
```diff
@@ -135,12 +139,25 @@ public String getServerUrl() {
   }
 
   /**
-   * @since 2.4
+   * @since X
```
Comment (luka5) [inline src/main/java/hudson/plugins/sonar/SonarInstallation.java]: TBD

**Comment by luka5** [comment]
@henryju Thanks for the feedback. Yes, I totally forgot about the integration tests. Could you please give me some hints on how I can execute these?

I also asked for hints about migrating the token as a credential in the [dev mailing list.](https://groups.google.com/forum/#!topic/jenkinsci-dev/gv8gyJ4XKmk)
I'll try to follow a similar approach as developed in [subversion-plugin](https://github.com/jenkinsci/subversion-plugin)

**Comment by henryju** [comment]
To run ITs, basic command is:
```
cd its
mvn -Djenkins.runtimeVersion=2.89.4 -Dsonar.runtimeVersion=7.2 clean verify
```
If you have troubles, let me know.

Also, based on Stephen feedback, I wonder if we should not introduce our own credential type (SonarToken ?)

**Comment by luka5** [comment]
closed in favor of #101

**Pull request closed without merging**